src/utils/__tests__/changelog.test.ts

```diff
@@ -0,0 +1,504 @@
+    '# Changelog\n\n   ## 1.0.0\n\n   this is a test\n',
+    '# Changelog\n\n   ## 2.0.0\n\n   - rewrote everything from scratch\n   - with multiple lines\n\n   ## 1.0.0\n\n   this is a test\n',
+  ],
+  [
+    'prepend with consistent padding with the rest (underlined)',
+    '# Changelog\n\n   1.0.0\n-----\n\n   this is a test\n',
+    '# Changelog\n\n   2.0.0\n-----\n\n   - rewrote everything from scratch\n   - with multiple lines\n\n   1.0.0\n-----\n\n   this is a test\n',
+  ],
+])('prependChangeset should %s', (_testName, markdown, expected) => {
+  expect(
+    prependChangeset(markdown, {
+      body: '- rewrote everything from scratch\n- with multiple lines',
+      name: '2.0.0',
+    })
+  ).toEqual(expected);
+});
+
+describe('generateChangesetFromGit', () => {
+  let mockClient: jest.Mock;
+  const makeCommitResponse = (commits: Record<string, string>[]) =>
+    mockClient.mockResolvedValueOnce({
+      repository: Object.fromEntries(
+        commits.map(({ hash, pr, milestone }: Record<string, string>) => [
+          `C${hash}`,
+          {
+            associatedPullRequests: {
+              nodes: pr
+                ? [
+                    {
+                      number: pr,
+                      milestone: milestone ? { number: milestone } : null,
+                    },
+                  ]
+                : [],
+            },
+          },
+        ])
+      ),
+    });
+
+  const mockGetChangesSince = getChangesSince as jest.MockedFunction<
+    typeof getChangesSince
+  >;
+  const dummyGit = {} as SimpleGit;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    mockClient = jest.fn();
+    (getGitHubClient as jest.MockedFunction<
+      typeof getGitHubClient
+      // @ts-ignore we only need to mock a subset
+    >).mockReturnValue({ graphql: mockClient });
+  });
+
+  it('should not fail on empty changeset', async () => {
+    mockGetChangesSince.mockResolvedValueOnce([]);
+    const changes = await generateChangesetFromGit(dummyGit, '1.0.0');
+    expect(changes).toBe('');
+  });
+
+  describe('standalone changes', () => {
+    it('should use short commit SHA for local commits', async () => {
+      mockGetChangesSince.mockResolvedValueOnce([
+        {
+          hash: 'abcdef1234567890',
+          message: 'Upgraded the kernel',
+          pr: null,
+        },
+      ]);
+      makeCommitResponse([]);
+      const changes = await generateChangesetFromGit(dummyGit, '1.0.0');
+      expect(changes).toMatchInlineSnapshot(`
+        "## Various fixes & improvements
+
+        - Upgraded the kernel (abcdef12)"
+      `);
+    });
+    it('should use short commit SHA for local commits w/o pull requests', async () => {
+      mockGetChangesSince.mockResolvedValueOnce([
+        {
+          hash: 'abcdef1234567890',
+          message: 'Upgraded the kernel',
+          pr: null,
+        },
+      ]);
+      makeCommitResponse([{ hash: 'abcdef1234567890' }]);
```

Comment:
This test is confusing. It's _almost_ the same as the one just before except for this line, but it's not obvious how this line means we are testing "w/o pull requests," because in both this and the previous test we have `pr: null` in the call to `mockResolvedValueOnce`.

Reply:
I infer from later tests that `mockResolvedValueOnce` simulates what's available locally while `makeCommitResponse` simulates what is available remotely. Not a big deal but maybe a comment or something could help to explain how the test suite is functioning.

Reply:
I'll also say that I have a personal preference for building test abstractions that highlight the variations in what is being tested. There's a lot of boilerplate repeated from test to test here that masks the meaning of each.

Reply:
But that would mean a solid refactor and I understand if you don't want to take that on. :)

Reply:
Brainstorm would be a `changesWith` wrapper that takes the variable elements and does the mock dancing, so that tests look like this:

```javascript
    it('should use short commit SHA for local commits', async () => {                                          
      var localCommits = [
        { 
          hash: 'abcdef1234567890',
          message: 'Upgraded the kernel',
          pr: null,
        },
      ];

      // remoteCommits and milestones would have defaults in changesWith but are here for illustration ...
      var remoteCommits = [];
      var milestones = [];

      expect(changesWith(localCommits, remoteCommits, milestones).toMatchInlineSnapshot(`
        "## Various fixes & improvements

        - Upgraded the kernel (abcdef12)"
      `);
    });
```

Then from here I guess we would get to parameterized testing. 😁

Reply:
Okay added a comment to alleviate the immediate confusion.

I agree with your suggestions but I'd defer that to a follow up PR as it indeed is a large refactor and I'd love to get this out first and better the tests later (as long as they cover the critical paths ofc).